### PR TITLE
Scrabble - Jackie & Leti (Ampers)

### PR DESCRIPTION
**Pull request by LetiTran**
# Scrabble
Congratulations! You're submitting your assignment.
## Comprehension Questions
Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? | One advantage is that both of us can contribute to the same project and both of us can have access to the most recent updates. We can also view a log of all the changes made by the other person. Git also helps manage conflicts.
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | We made score a class method so that we can directly access the method from other classes within the same Module Scrabble. 
Describe an example of a test you wrote for a _nominal case_. | We created a test to return the sum of scores of played words. We made our instance of a player play two words and in the assert statement we indicated that the score must equal the number we manually calculated.
Describe an example of a test you wrote for an _edge case_. | We wrote a test within scoring_spec in which if max score is tied and none of the words have a length of 7 the selected winning word would be the one with less characters.
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used select, max_by, and min_by in scoring.rb in the self.highest_score_from method. They were helpful to use in this method because it made our code more concise without having to use .each method.  
What was one method you and your pair used to debug code? | We inserted puts statements in places in which we thought the code was having issues. We also followed the stack trace to help us figure out where the errors were coming from.
Is there a specific piece of code you'd like feedback on? | Did we efficiently use max_by and min_by in scoring.rb in the self.highest_score_from method or are there other enumerable methods that are more appropriate for what we wanted to do.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | For us, the ping pong method allowed us to equally divide time spent driving and navigating. We consistently explained our code to each other which helped us understand what each part does. There was one part in wave 3 that we spent too much time trying to debug and we both think that if we were to do this again, we would have asked for help earlier. |

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -1,9 +1,95 @@
...
+      # If array with letters is empty, returns nil:
+      if letters_array.empty?
+        return nil
+      end
+
+
+      # Set the points for that word to initialize at zero and calculate the points of each word, adding them to the points variable:
+      points = 0
+
+      letters_array.each do |letter|
```
Comment (CheezItMan) [inline lib/scoring.rb]: This works, but is there a more compact way to do this, maybe with a hash?

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,95 @@
...
+# ________________HIGHEST SCORE METHOD___________________
+# Returns the word with highest score, from a given array of words:
+
+    def self.highest_score_from(array_of_words = [])
+      # Evaluate given words to find the one with highest score, according to the game rules. If there is a tie, it will populate a new array with the words in the tie and compare them to find the winner.
+
+      scoring_table = {}
+
+      # Set the scores of each word:
+      array_of_words.each {|word| scoring_table["#{word}"] = score(word)}
```
Comment (CheezItMan) [inline lib/scoring.rb]: You don't need the interpolation here:
```ruby
array_of_words.each {|word| scoring_table[word] = score(word)}
```

will work

File: specs/scoring_spec.rb [hunk 3]
```diff
@@ -42,21 +42,37 @@
...
 
     it 'returns the only word in a length-1 array' do
+      winning_word = Scrabble::Scoring.highest_score_from(["hotdog"])
+      winning_word.must_equal "hotdog"
+      winning_word.must_be_kind_of String
+
     end
 
     it 'returns the highest word if there are two words' do
+      winning_words = Scrabble::Scoring.highest_score_from(["dog", "academy"])
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:56]: Since order can potentially be important in the `highest_score_from` method repeating these tests by varying the order is useful. 

Something like:
```ruby
      winning_words = Scrabble::Scoring.highest_score_from(["academy", "dog"])
      winning_words.must_equal "academy"
```

File: lib/player.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    def draw_tiles(tile_bag)
+      @tiles = tile_bag.draw_tiles(7 - @tiles.count)
+      return @tiles
+    end
+
+
+# _________________PLAY METHOD____________________
+# Accepts a word as an argument and add it to the plays array.
+
+    def play(word)
```
Comment (CheezItMan) [inline lib/player.rb:29]: Read the description again.  This method is required to return `false` if they have already won prior to playing the word.

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        sum_of_scores += score
+      end
+      return sum_of_scores
+    end
+
+# _________________WON? METHOD____________________
+# Returns true if the toal score is more than 100, otherwise returns false.
+
+    def won?
+      total_score > 100 ? true : false
```
Comment (CheezItMan) [inline lib/player.rb:52]: Good ternary

File: specs/player_spec.rb [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    describe "#plays" do
+      it "Returns an array of words played:" do
+        player_1 = Scrabble::Player.new("Patrick")
+
+        player_1.plays.must_be_kind_of Array
+
+      end
+    end
+
+    describe "#play(word)" do
```
Comment (CheezItMan) [inline specs/player_spec.rb:31]: What about:
1.  Playing more than 1 word.
1.  Playing an invalid word
1.  Playing words **after** the player has already won!

File: specs/player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+  describe '#won?' do
+    it "Returns true if player has over 100 points" do
+      player_1 = Scrabble::Player.new("Patrick")
+      player_1.play("xxxxxxx")
+      player_1.won?.must_equal true
+
+    end
+
+    it "Returns false if player has over 100 points" do
```
Comment (CheezItMan) [inline specs/player_spec.rb]: `won?` should return `true` if the player has >= 100 points.  It's `play` that should return `false`

File: lib/tile_bag.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
+#***********************************************************
+# Class TileBag in Scrabble Module:
+
+module Scrabble
+  class TileBag
+
+    attr_reader :bag
+
+    def initialize
+      @bag = {
```
Comment (CheezItMan) [inline lib/tile_bag.rb]: There's a number of problems with this data structure.  Most pressing is that it gives each letter an **equal** chance of being drawn.  This is despite the fact that there are a **lot** more "E" tiles than "Z".

File: lib/tile_bag.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    # - delete the key of this letter from the tile bag if there are none available anymore.
+
+    def draw_tiles(num)
+      tiles_drawn = [ ]
+
+      num.times do
+        new_letter = @bag.keys.sample
+        tiles_drawn << new_letter
+        @bag[new_letter] -= 1
+        @bag.delete_if { |letter, quantity| quantity == 0 }
```
Comment (CheezItMan) [inline lib/tile_bag.rb:36]: Good that you're keeping the number of tiles from going negative.

File: specs/tile_bag_spec.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        sum_of_original_bag.must_equal 98 - 3
+        new_bag.length.must_equal 3
+        tile_bag.bag.values.wont_include :<, 0
+
+      end
+
+    end
+
+    describe '#tiles_remaining' do
+      it "returns number of tiles remaining in bag" do
```
Comment (CheezItMan) [inline specs/tile_bag_spec.rb:55]: You should also test to see what happens when I try to draw more tiles than are in the bag!

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Check
Both Teammates contributed to the codebase |  Check
Regular Commits with meaningful commit messages |  Good number of commits and good messages
**Scoring class** |
`score` method |  Check
Uses appropriate data structure to store the letter score |  Your way works, but there is a more compact way.
Appropriately handles edge cases |  Check
`highest_score_from` method |  Check
Appropriately handles edge cases |  Check
Test implementations for `highest_score_from`|  Yes, but see my note.  
**Player class** |
Uses an array to keep track of words played |  Check, but play isn't working properly
Uses existing code to create `highest_scoring_word` and `highest_word_score` |  Check
Returns `true` or `false` from `won?` method | Check
Tests edge cases on the `play(word)` |  **NOPE**, see my in-code notes on your tests.  You need to be **much** more through.  
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag |  See my comment on the data structure
`draw_tiles` method uses the parameter to draw a set of random tiles |  Check
`tiles_remaining` method returns a count |  Check
Tests `draw_tiles` edge cases |  You're missing the case where I try to draw 100 tiles.
Summary | Not bad you hit the learning goals.  You need to be a bit more through in your edge cases and see my note about the structure you use in `TileBag`.  However over all, nice work!
